src/components/tabs/Tabs.tsx

```diff
@@ -0,0 +1,112 @@
+import type { ComponentChildren } from 'preact';
+import { useEffect, useRef } from 'preact/hooks';
+import styles from './Tabs.module.css';
+import { useTabState } from './useTabState';
+
+const tabSlotKey = 'tab.' as const;
+const panelSlotKey = 'panel.' as const;
+
+type TabSlot = `${typeof tabSlotKey}${string}`;
+type PanelSlot = `${typeof panelSlotKey}${string}`;
+
+function isTabSlotEntry(entry: [string, ComponentChildren]): entry is [TabSlot, ComponentChildren] {
+	const [key] = entry;
+	return key.startsWith(tabSlotKey);
+}
+
+function isPanelSlotEntry(entry: [string, ComponentChildren]): entry is [PanelSlot, ComponentChildren] {
+	const [key] = entry;
+	return key.startsWith(panelSlotKey);
+}
+
+function getBaseKeyFromTab(slot: TabSlot) {
+	return slot.replace(new RegExp(`^${tabSlotKey}`), '')
+}
+
+function getBaseKeyFromPanel(slot: PanelSlot) {
+	return slot.replace(new RegExp(`^${panelSlotKey}`), '')
+}
+
+type Props = {
+	[key: TabSlot | PanelSlot]: ComponentChildren;
+	sharedStore?: string;
+}
+
+export default function Tabs({ sharedStore, ...slots }: Props) {
+	const tabs = Object.entries(slots).filter(isTabSlotEntry)
+	const panels = Object.entries(slots).filter(isPanelSlotEntry)
+	/** Used to focus next and previous tab on arrow key press */
+	const tabButtonRefs = useRef<Record<TabSlot, HTMLButtonElement | null>>({})
+	const scrollToTabRef = useRef<HTMLButtonElement | null>(null)
+
+	const firstPanelKey = panels[0]?.[0] ?? ''
+	const [curr, setCurr] = useTabState(getBaseKeyFromPanel(firstPanelKey), sharedStore)
+
+	function updateCurr(tabSlot: TabSlot, el: HTMLButtonElement | null) {
+		if (sharedStore) {
+			// Prevents scroll jumping due to layout shift
+			// from other tab views with the same store
+			scrollToTabRef.current = el
+		}
+		setCurr(getBaseKeyFromTab(tabSlot))
+	}
+
+	useEffect(() => {
+		if (scrollToTabRef.current) {
+			scrollToTabRef.current.scrollIntoView({ behavior: 'smooth' })
```

Comment:
There’s no way to sensibly scope this to _only_ Safari right? I’ve been playing around to try to avoid the scroll up, but can’t figure out a nicer behaviour. Without this works perfectly for me in FF & Chrome, so in a way this actually _worsens_ the UX there, but that may be unavoidable in our quest to make Safari not awful.